fixtures/no-bundle-import/src/index.test.ts

```diff
@@ -0,0 +1,41 @@
+import path from "path";
+import getPort from "get-port";
+import { describe, expect, test, beforeAll, afterAll } from "vitest";
+import { unstable_dev } from "../../../packages/wrangler/wrangler-dist/cli.js";
+import type { UnstableDevWorker } from "../../../packages/wrangler/wrangler-dist/cli.js";
+
+describe("Worker", () => {
+	let worker: UnstableDevWorker;
+
+	beforeAll(async () => {
+		worker = await unstable_dev(path.resolve(__dirname, "index.js"), {
+			bundle: false,
+			port: await getPort(),
+			experimental: { experimentalLocal: true },
+		});
+	}, 30_000);
+
+	afterAll(() => worker.stop());
+
+	test("module traversal results in correct response", async () => {
+		const resp = await worker.fetch();
+		const text = await resp.text();
+		expect(text).toMatchInlineSnapshot(
+			`"Hello Jane Smith and Hello John Smith"`
+		);
+	});
+
+	test("support for dynamic imports", async () => {
+		const resp = await worker.fetch("/dynamic");
+		const text = await resp.text();
+		expect(text).toMatchInlineSnapshot(`"dynamic"`);
+	});
+
+	test("no support for variable dynamic imports", async () => {
+		const resp = await worker.fetch("/dynamic-var");
+		const text = await resp.text();
+		expect(text).toMatchInlineSnapshot(
+			'"Error: No such module \\"dynamic-var.js\\"."'
```

Comment:
Yeah, this gets all the way to the runtime. Unfortunately, `esbuild` doesn't provide any hooks for dynamic variable imports—it seems to just ignore them (see https://esbuild.github.io/api/#non-analyzable-imports), so I'm not sure there's much we can do here. This _might_ be possible to support in future somehow with https://www.npmjs.com/package/es-module-lexer though

Reply:
If we can't find a way for `--no-bundle` to fail when dynamic imports can't be resolved, could this error message read something like:

"Error: The dynamically imported ES module \\"dynamic-var.js"\\ could not be found. Run `wrangler --dry-run --outdir --no-bundle` and verify that this module was generated when building your worker"

...such that we give people instructions on how to debug? (not sure if I've got the instructions quite right above, but you get the gist). How can we give people an error message they can act on and debug on their own?

Reply:
I'm not sure—this error comes from the runtime, so we may not want to call out Wrangler specifically. Maybe something like "Error: The dynamically imported ES module \"dynamic-var.js"\ could not be found. Make sure it was uploaded when you published your Worker"?

Reply:
I was going to ask about `[es-module-lexer](https://www.npmjs.com/package/es-module-lexer)`... could you please  do a lightweight exploration if implementing this feature via es-module-lexer would allow us to turn the runtime error into build time error?

If that was possible and there were no other major blockers, I'd be inclined to going down that path as runtime errors are really really nasty and we should avoid them whenever we can even if it means a little more work for us — it will save many times more work to developers that would otherwise scramble to deal with runtime errors.

Reply:
I've pushed up an additional commit that adds a build-time error for non string-literal dynamic imports. It uses `es-module-lexer` in addition to `esbuild`, and so has a slight performance cost over just using `esbuild` (but that should be minimal). They're both needed in order to be able to run Wrangler's module collection (which is implemented as an `esbuild` plugin), and to avoid re-implementing `esbuild`'s module resolution (reading `tsconfig`, for instance)

Reply:
I see. Interesting. How significant is the overhead? Do you have any estimates? <5ms for an average project, or more? Unless it's overhead measured in hundreds of ms, I think it's totally worth it. I am hoping for single or at most double digit `ms`...

Reply:
It's pretty minimal. `es-module-lexer` has benchmarks which estimate 5-10ms as an upper bound for parsing a MB of JS, but it will scale with the number of modules. For a large worker (let's say 25 modules at 200kb, which is at the 5MB limit for workers), `es-module-lexer` will probably add about 50ms of latency in the worst case (with a fairly pessimistic reading of their benchmarks).

Reply:
For an _average_ project (let's say 5 modules at 100kb), the overhead would be closer to ~5ms

Reply:
Great. Then this overhead is totally worth it. Thanks